# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
os.getcwd()

'/Users/ryanfisher/Documents/Springboard/Springboard/Guided Capstone 1'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
#os.listdir('data')
path= 'data'
os.listdir(path) 

['step2_output.csv', 'step3_output.csv']

In [23]:
df = pd.read_csv('data/step3_output.csv', header=0, index_col=0)
df

Name               Region       state  \
2                     Hilltop Ski Area               Alaska      Alaska   
4                  Sunrise Park Resort              Arizona     Arizona   
5        Yosemite Ski & Snowboard Area  Northern California  California   
8               Boreal Mountain Resort        Sierra Nevada  California   
9                          Dodge Ridge        Sierra Nevada  California   
..                                 ...                  ...         ...   
323                      Hogadon Basin              Wyoming     Wyoming   
326          Sleeping Giant Ski Resort              Wyoming     Wyoming   
327                   Snow King Resort              Wyoming     Wyoming   
328  Snowy Range Ski & Recreation Area              Wyoming     Wyoming   
329                White Pine Ski Area              Wyoming     Wyoming   

     summit_elev  vertical_drop  base_elev  trams  fastEight  fastSixes  \
2           2090            294       1796      0        0.0          0   
4          11100           1800       9200      0        0.0          0   
5           7800            600       7200      0        0.0          0   
8           7700            500       7200      0        0.0          0   
9           8200           1600       6600      0        0.0          0   
..           ...            ...        ...    ...        ...        ...   
323         8000            640       7400      0        0.0          0   
326         7428            810       6619      0        0.0          0   
327         7808           1571       6237      0        0.0          0   
328         9663            990       8798      0        0.0          0   
329         9500           1100       8400      0        0.0          0   

     fastQuads  ...  Snow Making_ac  daysOpenLastYear  yearsOpen  \
2            0  ...            30.0        150.000000       36.0   
4            1  ...            80.0        115.000000       49.0   
5            0  ...             0.0        110.000000       84.0   
8            1  ...           200.0        150.000000       54.0   
9            0  ...             0.0        115.103943       69.0   
..         ...  ...             ...               ...        ...   
323          0  ...            32.0        121.000000       61.0   
326          0  ...            18.0         61.000000       81.0   
327          0  ...           250.0        121.000000       80.0   
328          0  ...            30.0        131.000000       59.0   
329          0  ...             0.0        115.103943       81.0   

     averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
2               69.0     30.000000      34.00000         152.000000   
4              250.0     74.000000      78.00000         104.000000   
5              300.0     47.000000      47.00000         107.000000   
8              400.0     49.000000      64.16681         150.000000   
9              350.0     78.000000      78.00000         140.000000   
..               ...           ...           ...                ...   
323             80.0     48.000000      48.00000          95.000000   
326            310.0     42.000000      42.00000          77.000000   
327            300.0     59.000000      59.00000         123.000000   
328            250.0     49.000000      49.00000         120.053004   
329            150.0     57.916957      49.00000         120.053004   

     NightSkiing_ac  diff  clusters  
2              30.0   0.0         0  
4              80.0   0.0         2  
5               0.0   0.0         1  
8             200.0   0.0         1  
9               0.0   0.0         1  
..              ...   ...       ...  
323             0.0   0.0         1  
326             0.0   0.0         1  
327           110.0   0.0         1  
328             0.0   0.0         2  
329             0.0   0.0         2  

[176 rows x 29 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [28]:
#pd.get_dummies(df.state)
state_dummies = pd.get_dummies(df.state, prefix = 'state')

In [29]:
df = pd.concat([df, state_dummies], axis=1)
#df

In [46]:
df = df.drop(columns = 'state')

In [47]:
df = df.drop(columns = 'Region')

In [48]:
df

Name  summit_elev  vertical_drop  base_elev  \
2                     Hilltop Ski Area         2090            294       1796   
4                  Sunrise Park Resort        11100           1800       9200   
5        Yosemite Ski & Snowboard Area         7800            600       7200   
8               Boreal Mountain Resort         7700            500       7200   
9                          Dodge Ridge         8200           1600       6600   
..                                 ...          ...            ...        ...   
323                      Hogadon Basin         8000            640       7400   
326          Sleeping Giant Ski Resort         7428            810       6619   
327                   Snow King Resort         7808           1571       6237   
328  Snowy Range Ski & Recreation Area         9663            990       8798   
329                White Pine Ski Area         9500           1100       8400   

     trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
2        0        0.0          0          0     0       1  ...   
4        0        0.0          0          1     2       3  ...   
5        0        0.0          0          0     0       1  ...   
8        0        0.0          0          1     1       3  ...   
9        0        0.0          0          0     1       2  ...   
..     ...        ...        ...        ...   ...     ...  ...   
323      0        0.0          0          0     0       0  ...   
326      0        0.0          0          0     0       1  ...   
327      0        0.0          0          0     1       1  ...   
328      0        0.0          0          0     0       1  ...   
329      0        0.0          0          0     0       2  ...   

     state_Rhode Island  state_South Dakota  state_Tennessee  state_Utah  \
2                     0                   0                0           0   
4                     0                   0                0           0   
5                     0                   0                0           0   
8                     0                   0                0           0   
9                     0                   0                0           0   
..                  ...                 ...              ...         ...   
323                   0                   0                0           0   
326                   0                   0                0           0   
327                   0                   0                0           0   
328                   0                   0                0           0   
329                   0                   0                0           0   

     state_Vermont  state_Virginia  state_Washington  state_West Virginia  \
2                0               0                 0                    0   
4                0               0                 0                    0   
5                0               0                 0                    0   
8                0               0                 0                    0   
9                0               0                 0                    0   
..             ...             ...               ...                  ...   
323              0               0                 0                    0   
326              0               0                 0                    0   
327              0               0                 0                    0   
328              0               0                 0                    0   
329              0               0                 0                    0   

     state_Wisconsin  state_Wyoming  
2                  0              0  
4                  0              0  
5                  0              0  
8                  0              0  
9                  0              0  
..               ...            ...  
323                0              1  
326                0              1  
327                0              1  
328                0              1  
329         

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [164]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [ ]:
y

In [165]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.values.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [166]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [167]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [168]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
from sklearn.metrics import explained_variance_score,mean_squared_error
# You can plug y_test and y_pred into the functions to evaluate the model
explained_variance_score(y_test, y_pred)

-1.7724506528994996e+21

In [169]:
mean_absolute_error(y_test, y_pred)

87564509345.29755

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [170]:
model.intercept_

-21891127278.35317

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [171]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)

Coefficient
fastEight             1.004061e+13
total_chairs          7.631728e+12
diff                  5.339914e+12
fastSixes             5.276292e+12
double                4.189710e+12
surface               4.009472e+12
triple                3.422218e+12
quad                  1.948627e+12
trams                 1.756419e+12
fastQuads             1.628665e+12
state_New York        1.222806e+12
state_Michigan        1.075818e+12
state_California      9.711647e+11
state_New Hampshire   9.326499e+11
state_Pennsylvania    9.326499e+11
state_Wisconsin       9.326499e+11
state_Vermont         8.025645e+11
state_Minnesota       8.025645e+11
state_New Mexico      6.991662e+11
state_Idaho           6.991662e+11
state_Connecticut     6.401229e+11
state_Colorado        6.401229e+11
state_Maine           6.401229e+11
state_Wyoming         6.401229e+11
state_Massachusetts   5.742150e+11
state_Ohio            5.742150e+11
state_Washington      5.742150e+11
state_Montana         5.742150e+11
state_West Virginia   4.987283e+11
state_North Carolina  4.987283e+11
state_Illinois        4.987283e+11
state_Iowa            4.987283e+11
state_Utah            4.987283e+11
state_Oregon          4.987283e+11
state_Virginia        4.083852e+11
state_Missouri        4.083852e+11
state_Indiana         4.083852e+11
state_Tennessee       2.896005e+11
state_Rhode Island    2.896005e+11
state_Maryland        2.896005e+11
state_Alaska          2.896005e+11
state_South Dakota    2.896005e+11
state_Arizona         2.896005e+11
state_Nevada          2.896005e+11
state_New Jersey      8.131865e+06
summit_elev           2.472328e+01
base_elev             2.274275e+01
AdultWeekday          9.177002e+00
clusters              2.509277e+00
projectedDaysOpen     2.276611e+00
averageSnowfall       2.057129e+00
SkiableTerrain_ac     1.958984e+00
TerrainParks          1.842773e+00
vertical_drop         1.327125e+00
daysOpenLastYear      1.177368e+00
LongestRun_mi         9.554377e-01
Snow Making_ac        5.913086e-01
yearsOpen             4.682617e-01
NightSkiing_ac        3.923340e-01
Runs                  3.634644e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [139]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend','diff','clusters'], axis=1)
X = X.drop(X.loc[:,'state_Alaska':'state_Wyoming'],axis=1)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.loc[:,['AdultWeekend']] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [140]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.values.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [141]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [142]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)

Coefficient
AdultWeekday       1.032739e+01
base_elev          7.996105e+00
summit_elev        6.871495e+00
vertical_drop      2.756834e+00
TerrainParks       2.584444e+00
projectedDaysOpen  2.547779e+00
daysOpenLastYear   2.355922e+00
quad               2.206707e+00
SkiableTerrain_ac  2.096128e+00
LongestRun_mi      1.501800e+00
surface            1.142039e+00
Snow Making_ac     9.164105e-01
yearsOpen          8.319888e-01
Runs               7.420696e-01
double             5.823562e-01
NightSkiing_ac     4.761908e-01
total_chairs       3.651918e-01
fastQuads          1.877641e-01
triple             9.358340e-02
averageSnowfall    7.015950e-02
fastEight          2.486900e-14
fastSixes          2.398082e-14
trams              5.440093e-15

In [143]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.52638126536181
7.62769868405681


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [149]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend','summit_elev','base_elev','diff','clusters'], axis=1)
X = X.drop(X.loc[:,'state_Alaska':'state_Wyoming'],axis=1)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.loc[:,['AdultWeekend']] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [150]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.values.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [151]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [152]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.5285650304682641
7.587571911320586


In [148]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient',ascending=False)

Coefficient
AdultWeekday       1.023994e+01
vertical_drop      4.153243e+00
SkiableTerrain_ac  2.819264e+00
TerrainParks       2.577154e+00
projectedDaysOpen  2.202193e+00
quad               2.197168e+00
daysOpenLastYear   1.969581e+00
LongestRun_mi      1.439935e+00
Snow Making_ac     1.108255e+00
Runs               9.718713e-01
surface            9.505822e-01
yearsOpen          8.817270e-01
averageSnowfall    7.357153e-01
double             5.629623e-01
NightSkiing_ac     4.359878e-01
total_chairs       4.040474e-01
fastQuads          3.100456e-01
triple             7.309768e-02
fastEight          1.332268e-15
fastSixes          4.440892e-16
trams              0.000000e+00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -7.449240230538333e+19| 17951346601.787907 |-|
| Model 2. | 0.52638126536181 | 7.62769868405681 |'state'|
| Model 3. | 0.5285650304682641 | 7.587571911320586 |'state','summit_elev','base_elev'|

Model Selection:

In [0]:
Model 3 has the lowest MAE.